### PR TITLE
Init system property java.vm.specification.version via native

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Init system property `java.vm.specification.version` via native

There is a need to retrieve the system property `java.vm.specification.version` via `JVMTI API` in an early phase before Java code `java.lang.VersionProps.init(systemProperties)` running.
Removed the Java code (Java 12+) setting the property;
Get the property value via `JAVA_SPEC_VERSION` and set it within `vmprops.c:initializeSystemProperties()`;

Verified that the system property `java.vm.specification.version` has correct value.

closes: #5565 

Reviewer: @DanHeidinga 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: runtime/vm/vmprops.c [hunk 1]
```diff
@@ -572,35 +574,45 @@ initializeSystemProperties(J9JavaVM * vm)
...
-			classVersion = "55.0";
-			specificationVersion = "11";
-			break;
-		}
-		rc = addSystemProperty(vm, "java.class.version", classVersion, 0);
-		if (J9SYSPROP_ERROR_NONE != rc) {
+	if (JAVA_SPEC_VERSION == 8) {
+		specificationVersion = "1.8";
+	} else {
+		bufLen = j9str_printf(PORTLIB, NULL, 0, "%d", JAVA_SPEC_VERSION);
```
Comment (pshipton) [inline runtime/vm/vmprops.c]: Can you use the preprocessor for this?
https://stackoverflow.com/questions/240353/convert-a-preprocessor-token-to-a-string
https://github.com/eclipse/openj9/blob/abc3bd33fdef3079aa6e3f4a78b8dffec5b09e27/runtime/ddr/j9ddr.h#L33

File: runtime/vm/vmprops.c [hunk 2]
```diff
@@ -572,35 +574,45 @@ initializeSystemProperties(J9JavaVM * vm)
...
+	}
+	
+	if (JAVA_SPEC_VERSION < 12) {
+		/* Following system properties are defined via java.lang.VersionProps.init(systemProperties) within System.ensureProperties() */
+		const char *classVersion = NULL;
+		if (JAVA_SPEC_VERSION == 8) {
+			classVersion = "52.0";
+		} else {
+			/* Java 11 */
+			classVersion = "55.0";
```
Comment (pshipton) [inline runtime/vm/vmprops.c]: This should handle Java 12, 13, etc.

File: runtime/vm/vmprops.c [hunk 3]
```diff
@@ -572,35 +574,45 @@ initializeSystemProperties(J9JavaVM * vm)
...
+			}
+		} else {
+			rc = J9SYSPROP_ERROR_EMPTY_JAVA_SPEC_VERSION;
 			goto fail;
 		}
-
+	}
+	if (JAVA_SPEC_VERSION < 12) {
+		/* System property "java.specification.version" is defined via java.lang.VersionProps.init(systemProperties) within System.ensureProperties() */
 		rc = addSystemProperty(vm, "java.specification.version", specificationVersion, 0);
```
Comment (pshipton) [inline runtime/vm/vmprops.c]: I think we should be adding this even for Java 12 and up, even though VersionProps.init() sets it later. Then the value is available in JVMTI.

File: runtime/include/vendor_version.h [hunk 4]
```diff
@@ -49,17 +49,14 @@
 #define VENDOR_SHORT_NAME "OpenJ9"
 
 #define JAVA_VM_VENDOR "Eclipse OpenJ9"
+#define JAVA_VM_NAME "Eclipse OpenJ9 VM"
 
-#if JAVA_SPEC_VERSION < 12
```
Comment (pshipton) [inline runtime/include/vendor_version.h]: What is the purpose of this change?

File: runtime/vm/vmprops.c (hunk 1 above)
Comment (DanHeidinga) [inline runtime/vm/vmprops.c]: I'd like to avoid the allocation here if we can.  

Can you look at C preprocessor stringification - doing something like:
```C
#define STRINGIFY(x) #x 
specificationVersion = STRINGIFY(JAVA_SPEC_VERSION)

```

Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: Yeah, trying the hints from Pete's comment about `J9_STR(x)`.

Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: Copied the MACRO instead of including `"../ddr/j9ddr.h"`.

File: runtime/vm/vmprops.c (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: `JDK 12+` class versions are set by `java.lang.VersionProps.init(systemProperties)` within `System.ensureProperties()`.

File: runtime/include/vendor_version.h (hunk 4 above)
Comment (JasonFengJ9) [reply runtime/include/vendor_version.h]: `JAVA_VENDOR` & `JAVA_VENDOR_URL` was decorated by `#if JAVA_SPEC_VERSION < 12` in `vmprops.c`, now it is a runtime check instead hence these two need to be defined for `JAVA_SPEC_VERSION  12+` cases as well.

File: runtime/vm/vmprops.c (hunk 3 above)
Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: Done. Please have another look.

File: runtime/vm/vmprops.c [hunk 5]
```diff
@@ -572,39 +573,35 @@ initializeSystemProperties(J9JavaVM * vm)
...
 	}
 
-#if JAVA_SPEC_VERSION < 12
-	/* Following system properties are defined via java.lang.VersionProps.init(systemProperties) and following settings within System.ensureProperties() */
-	{
+	if (JAVA_SPEC_VERSION == 8) {
+		specificationVersion = "1.8";
+	} else {
+#define J9_STR_(x) #x
+#define J9_STR(x) J9_STR_(x)
```
Comment (DanHeidinga) [inline runtime/vm/vmprops.c]: We don't need two levels of this macro here as we're only working with a constant.
```c
#define J9_STR(x) #x 
```

File: runtime/vm/vmprops.c [hunk 6]
```diff
@@ -572,39 +573,35 @@ initializeSystemProperties(J9JavaVM * vm)
...
+		specificationVersion = J9_STR(JAVA_SPEC_VERSION);
+	}
+	rc = addSystemProperty(vm, "java.specification.version", specificationVersion, 0);
```
Comment (DanHeidinga) [inline runtime/vm/vmprops.c]: This is set by `VersionProps.init()`which the `java.vm.specification.version` is not.  To me, it makes sense for the VM to own setting the vm specific one, and the class libraries to own the other.

File: runtime/vm/vmprops.c (hunk 2 above)
Comment (pshipton) [reply runtime/vm/vmprops.c]: Since it used to be set in the VM, I think we should be adding this for Java 12 and up, even though VersionProps.init() sets it later. Then the value is available in JVMTI.

Comment (pshipton) [reply runtime/vm/vmprops.c]: Note it can be done automatically as 44 + JAVA_SPEC_VERSION

File: runtime/vm/vmprops.c (hunk 6 above)
Comment (pshipton) [reply runtime/vm/vmprops.c]: Since it used to be set in the VM prior to Java 12, I suggesting we should be adding this for Java 12 and up, even though VersionProps.init() sets it later. Then the value is available in JVMTI.

File: runtime/vm/vmprops.c (hunk 5 above)
Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: It appears it still requires two levels of macro [1] otherwise the system property value would be `JAVA_SPEC_VERSION`. 

[1] https://gcc.gnu.org/onlinedocs/gcc-4.8.5/cpp/Stringification.html

File: runtime/vm/vmprops.c (hunk 6 above)
Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: Agreed, updated PR to set `java.vm.specification.version` but not `java.specification.version` in native for `JDK12+`.

Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: Oops, missed Pete's comments above (no refresh).

File: runtime/vm/vmprops.c (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: Class version technically is owned by VM, set it in native for all Java versions. Please review again.

File: runtime/vm/vmprops.c [hunk 7]
```diff
@@ -572,39 +576,41 @@ initializeSystemProperties(J9JavaVM * vm)
...
+		goto fail;
+	}
+	rc = addSystemProperty(vm, "java.vm.specification.version", specificationVersion, 0);
+	if (J9SYSPROP_ERROR_NONE != rc) {
+		goto fail;
+	}
 
-		rc = addSystemProperty(vm, "java.vm.specification.version", specificationVersion, 0);
-		if (J9SYSPROP_ERROR_NONE != rc) {
+	bufLen = j9str_printf(PORTLIB, NULL, 0, "%d", classVersionValue);
```
Comment (pshipton) [inline runtime/vm/vmprops.c]: The class version is XX.0 not XX

File: runtime/vm/vmprops.c (hunk 6 above)
Comment (DanHeidinga) [reply runtime/vm/vmprops.c]: Doesn't this present a consistency issue?  The VM sets something early, its read in jvmti, and then VersionProps sets something subtly different at runtime. 

I would prefer to have the VM own the `*.vm.*` properties and let the class library own the others.  This seems to be inline with the design of VersionProps

File: runtime/vm/vmprops.c (hunk 5 above)
Comment (DanHeidinga) [reply runtime/vm/vmprops.c]: You're right.  I had misread the doc

**Comment by DanHeidinga** [comment]
Pete and I are providing opposite directions and I think that's making this a difficult PR to land.  Let's get the approach laid out and agreed to first.

My view is that properties set by `VersionProps.init()` [1] should NOT be pre-set by the VM as it may lead to consistency issues with subtle differences between values read early by jvmti and those read later.

I would prefer the VM own the `*.vm.*` properties and defer setting the ones from VersionProps to the java code.

This simplifies the VM code and removes the allocations currently being added from the startup paths.

@pshipton Can you comment on what you think the right approach here is?

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk12/blob/master/src/java.base/share/classes/java/lang/VersionProps.java.template

**Comment by pshipton** [comment]
My concern is that apps could see the change in behavior as a regression. Something that used to work up to jdk11, getting certain system properties via JVMTI, no longer works in jdk12 and later. Behavior is allowed to change on major version boundaries, and I do agree there is a risk of being inconsistent in the future if these properties stop following the expected pattern. It would be interesting to know what system properties Hotspot makes available via JVMTI.

**Comment by JasonFengJ9** [comment]
Built a `JVMTI` native and `jvmtiEnv->GetSystemProperty` within `Agent_OnLoad()` can return the system property `java.vm.specification.version` but not `java.specification.version` or `java.class.version`. 
So it appears `RI` native doesn't provide those properties set by `VersionProps.init()` at early phase as Dan expected.
Will update PR accordingly.

File: runtime/vm/vmprops.c (hunk 6 above)
Comment (JasonFengJ9) [reply runtime/vm/vmprops.c]: @DanHeidinga @pshipton the PR has been updated accordingly and ready for another reviewing.

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux jdk11

**Comment by DanHeidinga** [comment]
Jenkins test extended xlinux jdk8

**Comment by DanHeidinga** [comment]
Jenkins test sanity osx jdk12

**Comment by DanHeidinga** [comment]
@JasonFengJ9 I pushed some refactoring changes to your branch and kicked testing off on it.  Let me know if you have any concerns with the modifications.

> Built a JVMTI native and jvmtiEnv->GetSystemProperty within Agent_OnLoad() can return the system property java.vm.specification.version but not java.specification.version or java.class.version. 

Can you add a test for this under a separate PR?

**Comment by JasonFengJ9** [comment]
Thanks @DanHeidinga, the comments are much clearer.
In addition, I see the compile time check is preferred than run time check. It looks good to me.

>Can you add a test for this under a separate PR?

Will do.

**Comment by pshipton** [comment]
@DanHeidinga please cherry pick the commit(s) and create a PR for it against the v0.14.2-release branch.

**Comment by JasonFengJ9** [comment]
Probably I am going to do the cherry pick PR instead.
